Handle stateful properties in xfem moving interface

hugary1995 opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

This is a reopening of stale PR #15818.
This PR handles stateful properties when an XFEM cut is healed and re-cut.

closes #15817

hugary1995 committed:
handle stateful properties in xfem moving interface

store stateful properties before healing, and set them back after re-cut.

close #15817

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/630097/) on 84af071 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16278/site)

This comment will be updated on new commits.

jiangwen84 commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -18,6 +18,7 @@
 #include "libmesh/quadrature.h"
 
 #include "GeometricCutUserObject.h"
+#include "LevelSetCutUserObject.h"
```
Do we have to put `level set` in XFEM? I understand that you need `level set` to tell you the material belongs to which side. Can we make it not be too specific to `level set`? i.e. `mesh cut uo`.

hugary1995 replied:
Yeah I'll talk to you to see how we should design this.

hugary1995 replied:
Now this should be general enough to handle all geometric cuts. The related method is called `getConsistentSideBoolean` in the `GeometricCutUserObject` base class, which should be overridden by child classes. If it is not overridden AND healing is turned on AND there are stateful properties, it will throw an error.

hugary1995 commented:
Pinging @bwspenc .

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/630097/) on 84af071 : invalidated by @bwspenc

recuero commented on modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C:
```diff
@@ -25,8 +25,8 @@ ComputeFiniteStrainElasticStress::ComputeFiniteStrainElasticStress(
-    _rotation_total(declareProperty<RankTwoTensor>("rotation_total")),
-    _rotation_total_old(getMaterialPropertyOldByName<RankTwoTensor>("rotation_total")),
+    _rotation_total(declareProperty<RankTwoTensor>(_base_name + "rotation_total")),
+    _rotation_total_old(getMaterialPropertyOldByName<RankTwoTensor>(_base_name + "rotation_total")),
```
Necessary addition of `_base_name`

recuero replied:
Could you also add it to `ADComputeFinitStrainElasticStress`?

hugary1995 replied:
Oh I forgot that regular and AD are still in separate files here. I just did. Thanks for catching that.

bwspenc commented on modules/xfem/test/tests/moving_interface/moving_ad_bimaterial_finite_strain.i:
```diff
@@ -0,0 +1,271 @@
+# The global stress is determined by switching the stress based on level set values
+# The material interface is marked by a level set function
+# The two layer materials are glued together
+# This case is also meant to test for a bug in moving interfaces on displaced meshes
```
Since I assume the bug doesn't exist any more, I would reword this to not mention the bug, but say what the scenario that it tests to make sure works is.

bwspenc commented on modules/xfem/test/tests/moving_interface/moving_bimaterial_finite_strain.i:
```diff
@@ -0,0 +1,308 @@
+# The material interface is marked by a level set function
+# The two layer materials are glued together
+# This case is also meant to test for a bug in moving interfaces on displaced meshes
+# It should fail during the healing step of the 2nd timestep if the bug is present.
```
Same thing here about the bug.

hugary1995 commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1369,6 +1361,104 @@ XFEM::cutMeshWithEFA(NonlinearSystemBase & nl, AuxiliarySystem & aux)
       }
     }
+
+    // TODO: The 0 here is the thread ID.  Need to sort out how to do this correctly
```
I am not sure. This comment was there before I believe.

bwspenc replied:
I'd delete that comment then.

hugary1995 commented:
Can someone help me invalidate the failing tests due to conda?

loganharbour commented:
> Can someone help me invalidate the failing tests due to conda?

Looks good now.

bwspenc commented on modules/xfem/include/userobjects/GeometricCutUserObject.h:
```diff
@@ -182,6 +182,18 @@ class GeometricCutUserObject : public CrackFrontPointsProvider
   bool shouldHealMesh() const { return _heal_always; };
 
+  /**
+   * Get a consistent boolean telling which side the node belongs to relative to the cut
```
What do you mean by a 'consistent boolean'? I would suggest maybe renaming this class (or explaining what the consistent boolean is).

hugary1995 replied:
I want the function to return a boolean that is persistent across the entire simulation. The boolean itself should not contain any realistic information though. Can you teach me the English here, should it "consistent" or "persistent"?

bwspenc commented on modules/xfem/include/userobjects/LevelSetCutUserObject.h:
```diff
@@ -38,6 +38,8 @@ class LevelSetCutUserObject : public GeometricCutUserObject
   virtual const std::vector<Point>
   getCrackFrontPoints(unsigned int num_crack_front_points) const override;
 
+  virtual bool getConsistentSideBoolean(const Node * node) const override;
```
Add doxygen

hugary1995 replied:
This is overriding the function in the parent class. Do I still need to add doxygen here?

hugary1995 replied:
Well I added a doxygen anyways.

bwspenc reviewed (changes requested):
Really close!  Just a few little things to fix.

hugary1995 committed:
fix test prereq

hugary1995 committed:
xfem output files are causing issues

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/658070/) on e90f0db : invalidated by @bwspenc

Failure seems unrelated to this PR. Just checking.

bwspenc commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -2042,3 +2125,84 @@ XFEM::getNodeSolutionDofs(const Node * node, SystemBase & sys) const
+}
+
+bool
+XFEM::getElemConsistentSideBoolean(const Elem * parent_elem,
```
I think we still want to rename this method, right?

bwspenc replied:
`getElementSideOfInterface`?

bwspenc replied:
`getElementSideRelativeToInterface`?

hugary1995 replied:
I thought we are going to rename this method, along with the data structure and all the logics, in the next PR. Because later on we want to properly manage a vector of cut side IDs.

Anyways, I renamed it just for now.

hugary1995 committed:
rename

rwcarlsen commented on framework/include/materials/MaterialData.h:
```diff
@@ -79,6 +79,22 @@ class MaterialData
+   * element has no older properties
+   * @param side Side number (elemental material properties have this equal to zero)
+   */
+  void copy(const Elem & elem_to,
+            const HashMap<unsigned int, MaterialProperties> & props,
+            const HashMap<unsigned int, MaterialProperties> & props_old,
+            const HashMap<unsigned int, MaterialProperties> & props_older,
+            unsigned int side);
```
Just to satisfy my curiosity, why do you guys need to manually pass in the hashmaps here?  Why can't you just pass in a material data object like the other copy methods do?

rwcarlsen replied:
Or actually just pass in nothing (since this is a material data object) - and use the maps from the material data's mat prop storage object?

hugary1995 replied:
I think at the time we want to copy the material data, the elem_from no longer exists. 

At least that's why I didn't do what you suggested. Maybe you can help us think of a way around, so let me explain to you what we are trying to do:

1. First, there is this `element A`.
2. At some point, we want introduce a cut. So `element A` is cut with XFEM into `partial element B` and `partial element C`. `element A` is deleted.
3. Then, we want to update the location or orientation of the cut, so we first heal the cut. To do that, we delete `partial element B`, and heal `partial element C` to get `element D`.
4. Next, `element D` is re-cut with updated location and orientation of the cut, into `partial element E` and `partial element F`.

Then, obviously partial elements E and F lose the stateful information from partial elements B and C. This PR attempts to fix that. Before healing the cut in step 3, stateful properties from partial elements B and C are stored into a data structure we maintain in XFEM. After re-cutting the element in step 4, previously stored stateful data are copied into partial elements E and F using this new API in MaterialData we added.

With all that being said, it is still possible to avoid extracting and storing material properties from partial elements B and C if we can make them "live longer". In other words, if partial elements B and C are not deleted until we have created partial elements E and F, we can just use the old API to copy around the material properties without introducing this new API. I think @bwspenc can comment on whether this is possible without an overhaul of the current implementation of XFEM.

rwcarlsen replied:
Do steps 1 through 4 all happen on the same processor - or could there be a repartitioning or other such elem-moves between any of the steps?  Stateful material properties are stored in a separate data structure - not directly on/with the elements themselves.  So when you delete an element - its stateful property values should still be sitting inside the stateful prop storage data structures.  They are indexed by element pointers.  The problem you cited is that the element object may have been deleted - but it's previous pointer value should still work for indexing those data structures.  The real problem is that our existing APIs for copying/swapping, etc. take element references - when in actuality all they ever do is address the elem to get its underlying pointer/address.  I suspect that if we were to modify the existing APIs, you might not even need to copy around those material data objects - as long as the stateful properties had been synced to the main/sim-global storage loc by the time you want to do this copying.

bwspenc replied:
Thanks, @rwcarlsen, we'll do what you are suggesting!  That will considerably simplify things on our end.

rwcarlsen replied:
I suppose, however, that there is the risk that element deleting/creating could create a new element pointer with the same memory address as one of the previous elements - an address that was used to index a stateful property.  However, this would only be a problem if in the interrim, stateful properties associated with the new element reusing the old address were stored.  Seems really unlikely to affect you guys.  But this sort of thing is already a problem with our pointer-indexing as it is.  We really should use elem ids instead.  Someday we'll have time to redo all the material prop stuff.

hugary1995 committed:
use a simpler API

hugary1995 commented:
@rwcarlsen @bwspenc I just switched everything to use the API you suggested. But instead of changing the existing method using element references, I added a new method using pointers.

If we had been doing things correctly this should not break any test.

hugary1995 commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -355,6 +355,26 @@ MaterialPropertyStorage::copy(MaterialData & material_data,
+                              unsigned int side,
+                              unsigned int n_qpoints)
+{
+  initProps(material_data, *elem_to, side, n_qpoints);
```
Not sure how we should handle potential dereferencing errors here...

rwcarlsen replied:
Again - those functions all just need elem pointers - they never actually access the elem instance.

hugary1995 replied:
Ah you are right. Do you mind if I create another version of initProps that takes in element pointers?

hugary1995 committed:
add initProps that takes element pointer

rwcarlsen reviewed (changes requested):
All the original copy methods in MaterialData and MaterialPropertyStorage should just call the new functions you created - no need to duplicate all that code (i.e. `copy(const Elem & ...) {copy(&elem_to, &elem_from...);}`).  And since the initProps function is just private/internal - you should just delete the old version and make all calls use the new version.

hugary1995 committed:
remove duplicates
